### PR TITLE
Add TransportVersion to DiscoveryNode

### PR DESCRIPTION
**Pull request by thecoop**
Add a `getTransportVersion` to `DiscoveryNode`, to be propagated alongside node version, to represent the transport version used to communicate with the node.

Migrate an initial set of callsites away from `Version.transportVersion`

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/cluster/node/DiscoveryNode.java [hunk 1]
```diff
@@ -277,8 +375,9 @@ public DiscoveryNode(
      * @param address          the nodes transport address
      * @param attributes       node attributes
      * @param roles            node roles
-     * @param version          the version of the node
+     * @param nodeVersion      the version of the node
      */
+    @Deprecated(forRemoval = true)
```
Comment (thecoop) [inline server/src/main/java/org/elasticsearch/cluster/node/DiscoveryNode.java:380]: Eventually you will not be able to create a discoverynode without specifying node & transport versions explicitly

**Review by pgomulka — APPROVED** [review]
LGTM

File: server/src/main/java/org/elasticsearch/cluster/node/DiscoveryNode.java [hunk 2]
```diff
@@ -444,7 +604,10 @@ public void writeTo(StreamOutput out) throws IOException {
             o.writeString(role.roleNameAbbreviation());
             o.writeBoolean(role.canContainData());
         });
-        Version.writeVersion(version, out);
+        Version.writeVersion(nodeVersion, out);
+        if (out.getTransportVersion().onOrAfter(TRANSPORT_VERSION_VERSION)) {
```
Comment (thecoop) [inline server/src/main/java/org/elasticsearch/cluster/node/DiscoveryNode.java:608]: I *think* this recursive reference to transport version will Just Work, but if someone could confirm that would help

**Review by DaveCTurner — COMMENTED** [review]
Hmm this doesn't seem like the right approach to me. The transport version should be a property of the channel we're using to communicate with the remote node, which we should know from having completed a mutual handshake.

For instance:

- `JoinValidationService` and `PublicationTransportHandler` should collect up the channels we're intending to use for the transmission before we start to serialize anything.

- `Transport.Connection` should expose the transport version that the handshake negotiated

**Review by DaveCTurner — CHANGES_REQUESTED** [review]
Marking this as "request changes" until we finish discussing the content of my previous review.

**Comment by thecoop** [comment]
So how do we get from a `DiscoveryNode` instance to the transport version needed to communicate with that node?

File: server/src/main/java/org/elasticsearch/cluster/node/DiscoveryNode.java [hunk 3]
```diff
@@ -444,7 +604,10 @@ public void writeTo(StreamOutput out) throws IOException {
...
+            TransportVersion.writeVersion(transportVersion, out);
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/cluster/node/DiscoveryNode.java:609]: This means that every remote node will see the same `transportVersion` property for each `DiscoveryNode`, but they will not all be using the same transport protocol version.

**Comment by DaveCTurner** [comment]
> So how do we get from a DiscoveryNode instance to the transport version needed to communicate with that node?

We get a channel from the `TransportService` (i.e. the `ConnectionManager`) and ask it what transport version it negotiated.

**Comment by DaveCTurner** [comment]
Something _like_ that, yes. I'd prefer we only got the connection once in each case, using the same instance of `Transport.Connection` for retrieving the version and for sending the message.

**Comment by thecoop** [comment]
Closed in favour of #93822

**Pull request closed without merging**